agent/consul/discoverychain/gateway.go

```diff
@@ -0,0 +1,190 @@
+package discoverychain
+
+import (
+	"fmt"
+	"hash/crc32"
+	"sort"
+	"strconv"
+
+	"github.com/hashicorp/consul/agent/configentry"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// GatewayChainSynthesizer is used to synthesize a discovery chain for a
+// gateway from its configuration and multiple other discovery chains.
+type GatewayChainSynthesizer struct {
+	datacenter        string
+	gateway           *structs.APIGatewayConfigEntry
+	matchesByHostname map[string][]hostnameMatch
+	tcpRoutes         []structs.TCPRouteConfigEntry
+}
+
+type hostnameMatch struct {
+	match    structs.HTTPMatch
+	filters  structs.HTTPFilters
+	services []structs.HTTPService
+}
+
+// NewGatewayChainSynthesizer creates a new GatewayChainSynthesizer for the
+// given gateway and datacenter.
+func NewGatewayChainSynthesizer(datacenter string, gateway *structs.APIGatewayConfigEntry) *GatewayChainSynthesizer {
+	return &GatewayChainSynthesizer{
+		datacenter:        datacenter,
+		gateway:           gateway,
+		matchesByHostname: map[string][]hostnameMatch{},
+	}
+}
+
+// AddTCPRoute adds a TCPRoute to use in synthesizing a discovery chain
+func (l *GatewayChainSynthesizer) AddTCPRoute(route structs.TCPRouteConfigEntry) {
+	l.tcpRoutes = append(l.tcpRoutes, route)
+}
+
+// AddHTTPRoute takes a new route and flattens its rule matches out per hostname.
+// This is required since a single route can specify multiple hostnames, and a
+// single hostname can be specified in multiple routes. Routing for a given
+// hostname must behave based on the aggregate of all rules that apply to it.
+func (l *GatewayChainSynthesizer) AddHTTPRoute(route structs.HTTPRouteConfigEntry) {
+	for _, host := range route.Hostnames {
+		matches, ok := l.matchesByHostname[host]
+		if !ok {
+			matches = []hostnameMatch{}
+		}
+
+		for _, rule := range route.Rules {
+			// If a rule has no matches defined, add default match
+			if rule.Matches == nil {
+				rule.Matches = []structs.HTTPMatch{}
+			}
+			if len(rule.Matches) == 0 {
+				rule.Matches = []structs.HTTPMatch{{
+					Path: structs.HTTPPathMatch{
+						Match: structs.HTTPPathMatchPrefix,
+						Value: "/",
+					},
+				}}
+			}
+
+			// Add all matches for this rule to the list for this hostname
+			for _, match := range rule.Matches {
+				matches = append(matches, hostnameMatch{
+					match:    match,
+					filters:  rule.Filters,
+					services: rule.Services,
+				})
+			}
+		}
+
+		l.matchesByHostname[host] = matches
+	}
+}
+
+// Synthesize assembles a synthetic discovery chain from multiple other discovery chains
+// that have StartNodes that are referenced by routers or splitters in the entries for the
+// given CompileRequest.
+//
+// This is currently used to help API gateways masquarade as ingress gateways
```

Comment:
Just for my own understanding, is this stuff only here to bridge the gap between the prexisting gateway implementation and the new stuff we're making? When does it get called as part of the reconciliation loop.

Reply:
Yes! It allows us to avoid fully implementing xDS for API Gateway. Instead, we smoosh our configuration into an Ingress Gateway config and use the generated xDS from that.

In the reconciliation loop, there is the kick-off as any of the API Gateway configuration entries sees a change. The API Gateway as a whole is validated. Then a snapshot of its state is made.

That snapshot goes through this function which turns it into a snapshot of an Ingress Gateway. Then this sneaky Ingress Gateway snapshot is used to produce a discovery chain mapping the whole configuration out. This chain is then used to update xDS appropriately.

In the future, we'll get rid of discovery chains and have a different API for modifying xDS. So it didn't make sense to fully do that implementation in this cycle of work.